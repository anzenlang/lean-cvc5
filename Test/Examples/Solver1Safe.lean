import Test.Init

namespace cvc5.Safe.Test

def work : IO Unit := Smt.run! do
  Smt.setOption "produce-proofs" "true"
  Smt.setOption "produce-models" "true"

  Smt.setLogic "QF_LIA"

  let n1 ← Smt.declareConst "n1" Int
  let n2 ← Smt.declareConst "n2" Int

  let b ← Smt.declareConst "b" Bool

  let eq ← Smt.mkEqual n1 n2
  let neq ← Smt.mkNot eq
  let ite ← Smt.mkIte b eq neq

  println! "assert {ite}"
  Smt.assertFormula ite
  println! "assert {eq}"
  Smt.assertFormula eq

  println! "\nchecksat"
  Smt.checkSat
    (onSat := do
      println! "- confirmed `sat` result"
      let bVal ← Smt.getValue b
      println! "  getValue b := {bVal}"
      let intVals ← Smt.getValues #[n1, n2]
      println! "  getValues #[n1, n2] := {intVals}"
    )

  println! "\n\nreset\n\n"
  Smt.resetAssertions

  let imp1 ← Smt.mkImplies b eq
  let not_b ← Smt.mkNot b
  let imp2 ← Smt.mkImplies not_b neq

  println! "assert {imp1}"
  Smt.assertFormula imp1
  println! "assert {imp2}"
  Smt.assertFormula imp2
  println! "assert {eq}"
  Smt.assertFormula eq
  println! "assert {not_b}"
  Smt.assertFormula not_b

  println! "\nchecksat"
  Smt.checkSat
    (onUnsat := do
      println! "- confirmed `unsat` result"
      let proofs ← Smt.getProof
      println! "  proof :="
      for p in proofs do
        println! "    {p.getResult}"
    )


#eval work
