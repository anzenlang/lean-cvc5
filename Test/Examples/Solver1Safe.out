assert (let ((_let_1 (= n1 n2))) (ite b _let_1 (not _let_1)))
assert (= n1 n2)

checksat
- confirmed `sat` result
  getValue b := true
  getValues #[n1, n2] := #[0, 0]


reset


assert (=> b (= n1 n2))
assert (=> (not b) (not (= n1 n2)))
assert (= n1 n2)
assert (not b)

checksat
- confirmed `unsat` result
  proof :=
    (let ((_let_1 (not b))) (let ((_let_2 (= n1 n2))) (not (and (=> b _let_2) (=> _let_1 (not _let_2)) _let_2 _let_1))))
